### PR TITLE
Implement api v1 beacon block endpoints

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Implement API v1 beacon block endpoints.

## Fixed Issue(s)

#2761

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlock.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  public void handle(@NotNull final Context ctx) throws Exception {
+    final Map<String, String> pathParams = ctx.pathParamMap();
+    try {
+      final SafeFuture<Optional<SignedBeaconBlock>> future =
+          chainDataProvider.getBlock(pathParams.get(PARAM_BLOCK_ID));
+      handleOptionalResult(ctx, future, this::handleResult, SC_NOT_FOUND);
+    } catch (IllegalArgumentException ex) {
+      ctx.status(SC_BAD_REQUEST);
+      ctx.result(BadRequest.badRequest(jsonProvider, ex.getMessage()));
+    }
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlock.java]: this shouldn't be needed? if it's thrown during parse of the block id, that should be caught in the selector, and a BadRequestException should get thrown and it'll get handled by the generic exception handler.

if it is needed, the BlockSelectorFactory should be updated to try/catch in the same places the StateSelectorFactory has been made to catch for bytes32 parsing and numeric parsing.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockAttestations.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    final Map<String, String> pathParams = ctx.pathParamMap();
+    try {
+      final SafeFuture<Optional<List<Attestation>>> future =
+          chainDataProvider.getBlockAttestations(pathParams.get(PARAM_BLOCK_ID));
+      handleOptionalResult(ctx, future, this::handleResult, SC_NOT_FOUND);
+    } catch (IllegalArgumentException ex) {
+      ctx.status(SC_BAD_REQUEST);
+      ctx.result(BadRequest.badRequest(jsonProvider, ex.getMessage()));
+    }
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockAttestations.java]: shouldn't be needed

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockRoot.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  public void handle(@NotNull final Context ctx) throws Exception {
+    final Map<String, String> pathParams = ctx.pathParamMap();
+    try {
+      final SafeFuture<Optional<Root>> future =
+          chainDataProvider.getBlockRoot(pathParams.get(PARAM_BLOCK_ID));
+      handleOptionalResult(ctx, future, this::handleResult, SC_NOT_FOUND);
+    } catch (IllegalArgumentException ex) {
+      ctx.status(SC_BAD_REQUEST);
+      ctx.result(BadRequest.badRequest(jsonProvider, ex.getMessage()));
+    }
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockRoot.java]: shouldn't be needed

**Review by rolfyone — DISMISSED** [review]
just a couple of nits about exception handling location

File: data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java [hunk 4]
```diff
@@ -215,6 +216,33 @@ public GetForkResponse getForkInfo() {
...
+        .defaultBlockSelector(slotParameter)
+        .getSingleBlock()
+        .thenApply(
+            maybeBlock ->
+                maybeBlock.map(
+                    block ->
+                        block.getMessage().getBody().getAttestations().stream()
+                            .map(Attestation::new)
+                            .collect(toList())));
+  }
```
Comment (rolfyone) [inline data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java:245]: unit tests needed, fetch anything really, most of the provider tests have a mock ChainDataProvider, just checking responses are correct (eg. get the right block root)
I'm only expecting to see 1 test case, not extensive, as the defaultBlockSelectorFactory has tests already to show it's working. that'd pretty much cover the 'getBlock', but the others are pulling sections of data and returning.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockRoot.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      responses = {
+        @OpenApiResponse(
+            status = RES_OK,
+            content = @OpenApiContent(from = GetBlockRootResponse.class)),
+        @OpenApiResponse(status = RES_BAD_REQUEST),
+        @OpenApiResponse(status = RES_NOT_FOUND),
+        @OpenApiResponse(status = RES_INTERNAL_ERROR)
+      })
+  @Override
+  public void handle(@NotNull final Context ctx) throws Exception {
```
Comment (rolfyone) [inline data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockRoot.java:73]: it'd be nice to show the endpoints actually create data that can deserialize. the easiest way is an integration test, and checking for an 'OK' response that the data block can be read.
also should check that block root not found gives a 404 in an integration test

File: data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java (hunk 4 above)
Comment (cemozerr) [reply data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java:245]: Done.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlock.java (hunk 1 above)
Comment (cemozerr) [reply data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlock.java]: Removed.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockAttestations.java (hunk 2 above)
Comment (cemozerr) [reply data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockAttestations.java]: Removed.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockRoot.java (hunk 3 above)
Comment (cemozerr) [reply data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockRoot.java]: Removed.

File: data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockRoot.java (hunk 5 above)
Comment (cemozerr) [reply data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockRoot.java:73]: Done.

**Review by rolfyone — APPROVED** [review]
LGTM

**Pull request merged**